lib/reporters/html/partials/scripts.js

```diff
@@ -0,0 +1,89 @@
+'use strict';
+
+exports.scripts = () => {
+
+    return `<script>
+  headings = [];
+
+  onload = function () {
+      headings = document.querySelectorAll('h2');
+      reset();
+  };
+
+  onscroll = function (e) {
+      var heading = find(window.scrollY);
+      if (!heading) {
+        var current = document.querySelector('#menu a.active');
+        if (current) {
+          current.className = '';
+        }
+        return;
+      }
+      var links = document.querySelectorAll('#menu a')
+              , link;
+
+      for (var i = 0, len = links.length; i < len; ++i) {
```

Comment:
Property lookup for links.length each and every time through the loop is wasteful, as the value doesn't change during the execution of the loop.  By storing the value on len var we avoid subsequent lookups.

Reply:
Ref: https://www.oreilly.com/library/view/high-performance-javascript/9781449382308/ch04.html

Reply:
Array lookups are O(C) so either one should be equivalent. The advice in the link shared is specifically with regards to property lookups, which have the potential for additional costs due to needing to look up the prototype chain if local lookup fails.

Reply:
Oops, I think I miunderstood. The idea is to make it a constant-time lookup rather than looking up the property on an Array object. The point I believe still stands though bc JS engines optimize for these sorts of things. E.g. see https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/JIT_Optimization_Strategies.

Reply:
JIT compilation is _eventually_ performant.  First the JIT compiler needs to identify warm and hot code segments.  The warm code segments get sent to a baseline compiler and reused by the compiler where possible.  The hot code segments get sent to an optimizing compiler to be used until the assumptions under which the hot code was identified are invalidated at which point the optimized code is discarded (see https://blog.logrocket.com/how-javascript-works-optimizing-the-v8-compiler-for-efficiency/).

TL;DR there is a performance hit initially but indistinguishable for n sufficiently large, where n is the right hand of the for-loop conditional.